### PR TITLE
[Feature/Identity] User operations: create update delete

### PR DESCRIPTION
**Pull request by DarshitChanpura**
Signed-off-by: Darshit Chanpura <dchanp@amazon.com>

### Description
Adds C, U, D operations for internal user in an in-memory store.

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: sandbox/libs/authn/src/main/java/org/opensearch/authn/User.java [hunk 1]
```diff
@@ -7,10 +7,12 @@
 
 import com.fasterxml.jackson.annotation.JsonProperty;
 
+import java.security.Principal;
 import java.util.Collections;
+import java.util.List;
 import java.util.Map;
 
-public class User {
+public class User implements Subject {
```
[the remark this answers is not in the document]
Comment (DarshitChanpura) [reply sandbox/libs/authn/src/main/java/org/opensearch/authn/User.java:15]: I might be a bit confused here. In my understanding, a User is one of the many types of Subject, and InternalSubject and Subject were same. Let me know if my understanding is correct. 
Also, w.r.t. serialization/deserialization, isn't the PrincipalIdentifierToken the only serializable entity? Correct me if my understanding is off

[the remark this answers is not in the document]
Comment (peternied) [reply sandbox/libs/authn/src/main/java/org/opensearch/authn/User.java:15]: User represents the storage medium, Subject has functionality built onto it, and while the principal has strong alignment with the user, I would recommend we create another class to be a Subject that can swap out which User principal is pointed to.  Take a look at Shiros tutorial [1], its usage mirrors closely what I think we should implement here.  What do you think?

[1] https://shiro.apache.org/10-minute-tutorial.html

[the remark this answers is not in the document]
Comment (DarshitChanpura) [reply sandbox/libs/authn/src/main/java/org/opensearch/authn/User.java:15]: okay..makes sense... I will update this PR accordingly

**Comment by DarshitChanpura** [comment]
Precommit task `` fails with unused licenses:
<details>

```
* What went wrong:
Execution failed for task ':server:dependencyLicenses'.
> Unused sha files found: 
  lucene-queries-9.4.0.jar.sha1
  lucene-memory-9.4.0.jar.sha1
  lucene-analysis-common-9.4.0.jar.sha1
  lucene-grouping-9.4.0.jar.sha1
  lucene-highlighter-9.4.0.jar.sha1
  lucene-misc-9.4.0.jar.sha1
  lucene-spatial3d-9.4.0.jar.sha1
  lucene-spatial-extras-9.4.0.jar.sha1
  lucene-join-9.4.0.jar.sha1
  lucene-core-9.4.0.jar.sha1
  lucene-sandbox-9.4.0.jar.sha1
  lucene-queryparser-9.4.0.jar.sha1
  lucene-backward-codecs-9.4.0.jar.sha1
  lucene-suggest-9.4.0.jar.sha1
```
</details>

**Review by peternied — APPROVED** [review]
Lets merge this so we can build a Rest API action to create a user end to end.  Flip it to draft and we can see about merging

**Comment by DarshitChanpura** [comment]
We can merge this. I'll create a follow-up PR to add rest API for createUser functionality

**Pull request merged**